From: Hans-Joachim Picht <hpicht@redhat.com>
Date: Mon, 10 Dec 2007 14:04:27 +0100
Subject: [s390] zfcp: zfcp_erp_action_dismiss will ignore actions
Message-id: 20071210130427.GE29540@redhat.com
O-Subject: [RHEL5 U2 PATCH 5/8] s390 - zfcp: zfcp_erp_action_dismiss() used to ignore any actions in the ready list.
Bugzilla: 409091
RH-Acked-by: Pete Zaitcev <zaitcev@redhat.com>

Description
============

This patche changes zfcp_erp_action_dismiss() so that it dismisses actions
regardless of their list affiliation. The ERP thread is able to handle this.

It is important to kick the erp thread only for actions in the running list,
though, as an imbalance of wakeup signals would confuse the erp thread
otherwise.

Bugzilla
=========

BZ 409091
https://bugzilla.redhat.com/show_bug.cgi?id=409091

Upstream status of the patch:
=============================
Patch included in git as commit
d0076f7754dce07c7a1d752034561acadd99eafa

Test status:
============
Kernel with patch was built and successfully tested

Please ACK.

With best regards,

Hans

diff --git a/drivers/s390/scsi/zfcp_erp.c b/drivers/s390/scsi/zfcp_erp.c
index 55767ce..ed9bd32 100644
--- a/drivers/s390/scsi/zfcp_erp.c
+++ b/drivers/s390/scsi/zfcp_erp.c
@@ -981,7 +981,9 @@ static void zfcp_erp_action_dismiss(struct zfcp_erp_action *erp_action)
 	debug_text_event(adapter->erp_dbf, 2, "a_adis");
 	debug_event(adapter->erp_dbf, 2, &erp_action->action, sizeof (int));
 
-	zfcp_erp_async_handler_nolock(erp_action, ZFCP_STATUS_ERP_DISMISSED);
+	erp_action->status |= ZFCP_STATUS_ERP_DISMISSED;
+	if (zfcp_erp_action_exists(erp_action) == ZFCP_ERP_ACTION_RUNNING)
+		zfcp_erp_action_ready(erp_action);
 }
 
 int
